### PR TITLE
Adding tests and fixing up Firewall logic

### LINKED ISSUE #12: xFirewall: Get-DscConfiguration fails
When xFirewall is used in a DSC configuration, Get-DscConfiguration fails with the error:
"Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.Collections.IList'."

### PR DESCRIPTION
**Pull request by tysonjhayes**
This fixes #12 and adds some integration testing and improves unit testing for xFirewall which helps #14.

I'd appreciate some extra testing on this before I merge as this is likely going to be a breaking change as I changed some of the schema to remove some unneeded parameters and changed some of the parameters to reduce code paths.

### CODE REVIEW

**Comment by tysonjhayes** [comment]
Alright if anyone wants to code review this I'd be really appreciative. I've had a couple of people on this side take a look, but I'd love someone from the community to take a look as well.

**Comment by Dan1el42** [comment]
Thanks @tysonjhayes Great work. Just a few things myself and the script analyzer have noticed.

Unit/MSFT_xFirewall.Tests.ps1
1. Please expand the "select" alias in line 51, 112, 121 and 128 to "Select-Object".
2. Please consider to rename the variable $firewall to $rule. For example $firewall.DisplayName does not seem to make sense without reading the line "$firewall = Get-NetFirewallRule | ..." everytime.
3. Please consider to add a line break before "select/Select-Object" on line 112, 121, 128, 137 and before "Should Be" in all "It" blocks of Describe "Get-FirewallRuleProperty" and indent appropriately to stay within the 100 characters guideline.
````
$firewall = Get-NetFirewallRule | Where-Object {$_.DisplayName -ne $null} | 
 Select-Object -First 1
````
4. Please consider to use splatting for "$result = Test-TargetResource ..." on line 115, 122 and 130 to stay within 100 characters.
5. Please consider to remove the commented -DisplayName $firewall.Name from line 157 and 168.
6. Please consider to use a script variable for -Destination of the Copy-Item cmdlet in line 236 to shorten the line under 100 characters.

Unit/MSFT_xIPAddress.Tests.ps1
1. Please consider to add a line break before "Should Throw ..." on line 71 to shorten the line under 100 characters and replace double quotes with single quotes.
2. Please consider to use a script variable for -Destination of the Copy-Item cmdlet in line 236 to shorten the line under 100 characters.

P.S. ISESteroids has a great feature called "Global Guide" under the vertical guideline toggle to draw a line at 100 character column position.

**Comment by tysonjhayes** [comment]
Thanks @Dan1el42 your suggestions seem pretty reasonable I'll work on getting them implemented and tag you once I've got them committed.

**Comment by tysonjhayes** [comment]
OK - that was way more fighting with AppVeyor then I expected. I'm going to add the Integration tests as a seperate PR as it's being dumb and I'd like to get these changes added to actually fix the pr I started this for.  @PlagueHO as you just added some tests could you give this a pass? I'm looking over your code as well to see if it makes sense to merge these together somehow. @Dan1el42 this should more or less modify the code as you suggested.

I'll wait a while before merging for anymore comments.

File: DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1 [hunk 1]
```diff
@@ -325,154 +257,46 @@ function Test-TargetResource
...
-        Write-Verbose "Test-RuleHasProperties: Name property value - $FirewallRule.Name does not match desired state - $Name"
-
-        $desiredConfigurationMatch = $false
-    }
-
-    if ($Access -and ($FirewallRule.Action -ne $Access))
-    {
-        Write-Verbose "Test-RuleHasProperties: Access property value - $($FirewallRule.Action) does not match desired state - $Access"
-
+        Write-Verbose "$($MyInvocation.MyCommand): Name property value - $FirewallRule.Name does not match desired state - $Name"
```
Comment (PlagueHO) [inline DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1]: I think $FirewallRule.Name needs to be enclosed in parentheses - e.g. $($FirewallRule.Name)

File: DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1 [hunk 2]
```diff
@@ -325,154 +257,46 @@ function Test-TargetResource
...
         $desiredConfigurationMatch = $false
     }
 
-    if ($State -and ($FirewallRule.Enabled.ToString() -eq ("Enabled" -ne $State)))
+    if ($Enabled -and ($FirewallRule.Enabled.ToString() -eq $Enabled))
     {
-        Write-Verbose "Test-RuleHasProperties: State property value - $FirewallRule.Enabled.ToString() does not match desired state - $State"
-
+        Write-Verbose "$($MyInvocation.MyCommand): State property value - $FirewallRule.Enabled.ToString() does not match desired state - $Enabled"
```
Comment (PlagueHO) [inline DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1]: I think $FirewallRule.Enabled.ToString() needs to be enclosed in parentheses - e.g. $($FirewallRule.Enabled.ToString())

File: DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1 [hunk 3]
```diff
@@ -480,26 +304,22 @@ function Test-RuleHasProperties
...
+            Write-Verbose "$($MyInvocation.MyCommand): Profile property value - '$networkProfileinRule' does not match desired state - '$Profile'"
             $desiredConfigurationMatch = $false
         }
     }
 
     if ($Direction -and ($FirewallRule.Direction -ne $Direction))
     {
-        Write-Verbose "Test-RuleHasProperties: Direction property value - $FirewallRule.Direction does not match desired state - $Direction"
-
+        Write-Verbose "$($MyInvocation.MyCommand): Direction property value - $FirewallRule.Direction does not match desired state - $Direction"
```
Comment (PlagueHO) [inline DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1]: I think $FirewallRule.Direction needs parentheses too :)

File: DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1 [hunk 4]
```diff
@@ -536,68 +356,62 @@ function Test-RuleHasProperties
...
+            Write-Verbose "$($MyInvocation.MyCommand): LocalPort property value - '$localPortInRule' does not match desired state - '$LocalPort'"
             $desiredConfigurationMatch = $false
         }
     }
 
     if ($Protocol -and ($properties.PortFilters.Protocol -ne $Protocol))
     {
-        Write-Verbose "Test-RuleHasProperties: Protocol property value - $properties.PortFilters.Protocol does not match desired state - $Protocol"
-
+        Write-Verbose "$($MyInvocation.MyCommand): Protocol property value - $properties.PortFilters.Protocol does not match desired state - $Protocol"
```
Comment (PlagueHO) [inline DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1]: Ditto :)

File: DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1 [hunk 5]
```diff
@@ -536,68 +356,62 @@ function Test-RuleHasProperties
...
         $desiredConfigurationMatch = $false
     }
 
     if ($Description -and ($FirewallRule.Description -ne $Description))
     {
-        Write-Verbose "Test-RuleHasProperties: Description property value - $FirewallRule.Description does not match desired state - $Description"
-
+        Write-Verbose "$($MyInvocation.MyCommand): Description property value - $FirewallRule.Description does not match desired state - $Description"
```
Comment (PlagueHO) [inline DSCResources/MSFT_xFirewall/MSFT_xFirewall.psm1]: Here too!

File: Examples/Sample_xFirewall_AddFirewallRule.ps1 [hunk 6]
```diff
@@ -18,7 +18,7 @@ configuration Sample_xFirewall_AddFirewallRule
             DisplayGroup          = "NotePad Firewall Rule Group"
             Ensure                = "Present"
             Access                = "Allow"
```
Comment (PlagueHO) [inline Examples/Sample_xFirewall_AddFirewallRule.ps1]: Access parameter has been removed from Resource.

File: Examples/Sample_xFirewall_AddFirewallRuleToExistingGroup.ps1 [hunk 7]
```diff
@@ -27,7 +27,7 @@ configuration Sample_xFirewall_AddFirewallRuleToExistingGroup
             DisplayGroup          = "My Firewall Rule Group"
             Ensure                = "Present"
             Access                = "Allow"
```
Comment (PlagueHO) [inline Examples/Sample_xFirewall_AddFirewallRuleToExistingGroup.ps1]: Access parameter has been removed from resource.

File: README.md [hunk 8]
```diff
@@ -37,12 +37,12 @@ Please check out common DSC Resources [contributing guidelines](https://github.c
 
 ### xFirewall
 
-* **Name**: Name of the firewall rule 
+* **Name**: Name of the firewall rule
 * **DisplayName**: Localized, user-facing name of the firewall rule being created .
 * **DisplayGroup**: Name of the firewall group where we want to put the firewall rules.
 * **Ensure**: Ensure that the firewall rule is Present or Absent.
 * **Access**: Permit or Block the supplied configuration.
```
Comment (PlagueHO) [inline README.md]: Access has been removed from resource.

File: README.md [hunk 9]
```diff
@@ -243,25 +248,25 @@ Configuration Sample_xDefaultGatewayAddress_Remove
...
+
+    Import-DSCResource -ModuleName xNetworking
+
+    Node $NodeName
+    {
+        xFirewall Firewall
+        {
+            Name                  = "MyAppFirewallRule"
+            ApplicationPath       = "c:\windows\system32\MyApp.exe"
+            Access                = "Allow"
```
Comment (PlagueHO) [inline README.md]: Access has been removed from resource.

File: README.md [hunk 10]
```diff
@@ -270,36 +275,36 @@ Configuration Add_FirewallRule
...
+    Import-DSCResource -ModuleName xNetworking
+
+    Node $NodeName
+    {
+        xFirewall Firewall
+        {
+            Name                  = "MyFirewallRule"
+            DisplayName           = "My Firewall Rule"
+            DisplayGroup          = "My Firewall Rule Group"
+            Access                = "Allow"
```
Comment (PlagueHO) [inline README.md]: Access has been removed from resource.

**Comment by PlagueHO** [comment]
I think the changes are looking great (I was delighted about the change from state to enabled :+1:  ). This also makes the issue #25 a bit easier to manage.

I noted a few minor things and put line notes in - was that the best way of providing feed back or would it have been better to just put it all in a single comment? Just for future reference.

But I do have one question (and I might be being really dumb so forgive me), but it seems that the Access (Allow, Block) property has been removed from the resource (mostly). That seems to mean that whenever a rule is created it will create it with a default Action of Allow. So I can't see how a Block rule could be created? I'm thinking we still need an Access rule parameter (although it would be better if it was called Action so that it matched the *-NetFirewallRule CmdLet parameters) to enable this.

If Access is meant to be removed from the resource then I did note a few places it was still referred to in the examples/docs.

As for the Pester Tests, your ones look pretty similar to the ones I created. I can see that with the changes you've made I can actually get rid of a bunch of mine. I think we go with what you've got and then see if there is anything I added that was worth moving into yours one.

**Comment by tysonjhayes** [comment]
I think the problem with working on this on and off for the past month is that I got my wires crossed. I think I was trying to remove state, not access, however you make a good point that Access is really Action, so we should probably rename it as such so its clearer on what its doing.  Thanks for the catch @PlagueHO !

**Comment by PlagueHO** [comment]
Hehe that makes sense :+1:

**Comment by tysonjhayes** [comment]
OK we're back to if everything looks good I'll merge it, given this has been open for 30 days if I get a go ahead from anyone I'll merge it and hopefully not break everything! :laughing:

**Comment by Dan1el42** [comment]
Looks great to me. Thanks so much for putting the hours in to improve the test coverage of the xNetworking resources.

P.S. Only very few lines are longer than the 100 character rule but I think we should get this merged regardless and resolve that later with a quick PR.

**Comment by tysonjhayes** [comment]
@Dan1el42 d'oh! Thought I got all of those. Yeah I'm going to merge it, then lets get a PR to fix it up from here. Thanks for the review!

**Comment by PlagueHO** [comment]
Awesome stuff! Looks great to me :+1:

**Pull request merged**